test/SwiftSyntax/DiagnosticTest.swift

```diff
@@ -58,4 +70,36 @@ Diagnostics.test("DiagnosticEmission") {
   expectEqual(fixIt.text, " != 0")
 }
 
+Diagnostics.test("SourceLocations") {
+  let engine = DiagnosticEngine()
+  engine.addConsumer(PrintingDiagnosticConsumer())
+  let url = getInput("diagnostics.swift")
+
+  class Visitor: SyntaxVisitor {
+    let url: URL
+    let engine: DiagnosticEngine
+    init(url: URL, engine: DiagnosticEngine) {
+      self.url = url
+      self.engine = engine
+    }
+    override func visit(_ function: FunctionDeclSyntax) {
+      let startLoc = function.identifier.startLocation(in: url)
+      engine.diagnose(.badFunction(function.identifier), location: startLoc) {
```

Comment:
can we issue diags to the end of the function decl to test the end location too?